### ISSUE TITLE
Add support for bumpy array

### ISSUE THREAD

**Issue by dvarrazzo** [psycopg/psycopg#336]
Natural prosecution of #192.

Binary dumping/loading would be pretty efficient, thanks to `ndarray.tobytes`. 

It requires a dumper implementing upgrade(), because different dtypes should be dumped as different oid and with different format.

**Comment by dvarrazzo** [comment]
I have actually received very little request for this feature. Therefor I'm thinking about not blocking 3.2 for this feature, which would require some non-trivial amount of work to be implemented.

**Comment by JeffreyEarly** [comment]
This might be exactly what we are looking for. We have a query that returns, e.g., (x,y,z) values that we then need to do some computations on with numpy. This query and subsequent computation gets repeated a lot, and needs to be fast. So, we are interested in getting the data back from the database into a numpy array with minimal overhead. What's currently the best way to do that? Would this take care of what we need?

Thanks,
Jeffrey

**Comment by dvarrazzo** [comment]
@JeffreyEarly At the moment, psycopg can [store and retrieve arrays based on nested lists](https://www.psycopg.org/psycopg3/docs/basic/adapt.html#lists-adaptation). So there must be an intermediate step where you transform your array to lists in Python, and store them to the database, and, when you fetch from the database, you convert the nested lists you obtained into a numpy array.

You can actually create your own numpy dumper and loader which could do this intermediate step automatically. You can follow [the XML example](https://www.psycopg.org/psycopg3/docs/advanced/adapt.html#writing-a-custom-adapter-xml) in the documentation and gather how to do it.

If we add numpy array support directly to psycopg, it might result in a more optimized operation, which might be somewhat faster and taking less memory than passing via the intermediate list representation. But, before that, you could build a simple adapters prototype as described so you can verify if it fulfills your needs.

**Comment by j-carson** [comment]
I'm confused about your comment that a numpy binary dumper needs to "set the oid" -- the actual logic to dump a numpy float32 array to a postgres float4 binary dump should just be (in numpy 2.0 land):
 
```python
my_array.view(my_array.dtype.newbyteorder(">")).data
```

The confusing part is that the dumper doesn't seem to take the destination type as an argument or as a return val so I don't know whether float4 is what the dumper wants, or how to tell psycopg that the data was dumped as a float4? 

I currently dump numpy arrays as strings in this fashion:

```python
class NParrayDumper(psycopg.adapt.Dumper):

    format = psycopg.pq.Format.TEXT

    def dump(self, obj: np.ndarray) -> bytes:
        """Change an array to string in postgres copy format.

        Parameters
        ----------
        obj: np.ndarray
            The data to load as an array in sql

        Returns
        -------
        result: bytes
            Postgres array boundaries are curly braces, and commas separate the values
            Example one dimensional array:"{1,2,3}", example two-dimensional array:
            "{{1,2,3},{4,5,6}}"

        Note
        ----
        Throws an exception if the array has more than two dimensions.
        """
        ndim = len(obj.shape)

        if ndim == 1:
            return f'{{{",".join([str(item) for item in obj])}}}'.encode()

        elif ndim == 2:
            # This fstring in an f-string with escaped-curlies is ugly-but-performant
            # End goal is: {{item,item,item},{item,item,item}}
            return f"""{{{','.join([f"{{{','.join([str(item) for item in row])}}}" for row in obj])}}}""".encode()  # noqa

        raise ValueError(
            f"ndarray_dumper: must be one or two dimensional array, got {obj.shape}"
        )
```

and is registered as:
```
curs.adapters.register_dumper(np.ndarray, NParrayDumper)
```

**Comment by dvarrazzo** [comment]
> I'm confused about your comment that a numpy binary dumper needs to "set the oid"

Psycopg instantiates a dumper for every type of the parameters passed to a query. if you pass the query a (10, 20, 'hello'), it will instantiate one `IntDumper` and one `StrDumper`, and will ask the first to convert the first two values, the second to convert the third value.

However, sometimes, just looking at the type is not enough, because sometimes the same Python type could be better represented by different postgres types. For instance a python list could be a postgres int array or string array, according to the elements; a datetime could be a timestamp or a timestamptz according to whether it has a tzinfo.

So the dumpers have an upgrade mechanism to deal with Python types that may have more than a Postgres type representing them. This is described in the [`Dumper.get_key()` and `upgrade()`](https://www.psycopg.org/psycopg3/docs/api/abc.html#psycopg.abc.Dumper.get_key) documentation.

This is the case in which numpy is, because different arrays have the same python types:

```python
import numpy as np
a1 = np.zeros((2,2), dtype=np.float32)
a2 = np.zeros((2,2), dtype=np.int64)
type(a1) is type(a2)
True
```

But for postgres they are two different types:

```
piro=# select pg_typeof('{}'::float[])::oid;
 pg_typeof 
-----------
      1022
(1 row)

piro=# select pg_typeof('{}'::bigint[])::oid;
 pg_typeof 
-----------
      1016
(1 row)
```

The array dumper implements such a logic, and it's actually much more complex because the list doesn't have a dtype so we need to find a representative value, avoid the None, etc. Something that has a closer implementation is with [the datetime dumper](https://github.com/psycopg/psycopg/blob/fcbe775da0aab8591006918e6c09580241462966/psycopg/psycopg/types/datetime.py#L133-L164).

I think that there should be a hierarchy of dumpers, where the base class has the implementation of the dumping procedure but doesn't know any OID. It should have an `upgrade()` method that would return instead the right subclass.

> The confusing part is that the dumper doesn't seem to take the destination type as an argument or as a return val so I don't know whether float4 is what the dumper wants, or how to tell psycopg that the data was dumped as a float4?

These subclass can specify the type using their `oid` attribute. Sketching...

```python
# Concrete classes, they know how to dump a specific dtype
# They could inherit from a common class if useful, inherit from ArrayDumper if it makes sense...

class Float32ArrayDumper:
    oid = psycopg.adapters.types["float4"].array_oid
    def dump(self): ...

class Int64ArrayDumper:
    oid = psycopg.adapters.types["int8"].array_oid
    def dump(self): ...


# This is a dispatch class that takes the input parameter and figures out the concrete class 

class ArrayDumper(Dumper):
    # oid not set

    _dumper_by_dtype = {
        np.dtype("float32"): Float32ArrayDumper,
        np.dtype("float32"): Float32ArrayDumper,
        np.dtype("int64"): Int64ArrayDumper,
    }

    def dump(self):
        # base implementation, or throw a NotImplementedError.

    def get_key(self, obj):
        # return something to represent this specialization, for instance
        return (type(self), type(obj.dtype))

    def upgrade(self, obj):
        # return a proper dumper
        return self._dumper_by_dtype[obj.dtype](obj)


# Finally, we map the type to the dumper (the name of the type is enough).
# This makes so that when an array is passed to a query the ArrayDumper is instantiated,
# then at the right time it will dispatch to a concrete dumper which will know how to dump
# any array passed to the query.

def register_default_adapters(context: AdaptContext) -> None:
    adapters = context.adapters
    adapters.register_dumper("numpy.ndarray", ArrayDumper)
```

> the actual logic to dump a numpy float32 array to a postgres float4 binary dump should just be (in numpy 2.0 land):

That, or something similar, with the addition of a postgres array header that can also be seen in the array dumper implementation module.

**Comment by dvarrazzo** [comment]
I believe you found it in [`psycopg.abc.Dumper.get_key`](https://www.psycopg.org/psycopg3/docs/api/abc.html#psycopg.abc.Dumper.get_key), despite the autodoc glitch you found.

A dumper key is either a type or a tuple of dumper keys (so nested tuples of types). It doesn't really matter what it is: as long as `get_key()` and `upgrade()` agree on how to use it.

Unless I'm mistaken, as described above, `(type(self), type(obj.dtype))` seems a good key.

**Comment by j-carson** [comment]
Here is my "upgrade" function if it is helpful to anyone else

```python
import numpy as np
from functools import cache
import psycopg
from psycopg.types.array import ArrayBinaryLoader
import pytest
from math import prod

SUPPORTED_NUMPY_TYPES = {np.float32, np.float64, np.int16, np.int32, np.int64}
from psycopg.adapt import Dumper
from psycopg.pq import Format


@cache
def ndarray_binary_dumper(dtype: type) -> Dumper:
    """Generate a dumper to dump a numpy array of type
    "dtype" to binary (which will be a singleton
    due to @cache decorator)"""

    numpy_type_spec = {
        np.float32: ">f4",
        np.float64: ">f8",
        np.int16: ">i2",
        np.int32: ">i4",
        np.int64: ">i8",
    }[dtype]
    numpy_oid = {
        np.float32: psycopg.adapters.types["float4"],
        np.float64: psycopg.adapters.types["float8"],
        np.int16: psycopg.adapters.types["int2"],
        np.int32: psycopg.adapters.types["int4"],
        np.int64: psycopg.adapters.types["int8"],
    }[dtype]

    class TypeSpecificDumper(Dumper):
        oid = numpy_oid.array_oid
        format = Format.BINARY

        def __init__(self, cls: type) -> None:
            super().__init__(cls)

        def dump(self, obj: np.ndarray) -> bytes:
            item_size = obj.itemsize
            n_items = prod(obj.shape)
            element_oid = numpy_oid.oid
            # Data type that adds a size attribute to each item of 'obj'
            sized_item_dtype = np.dtype(
                [
                    ("item_size", ">u4"),
                    ("item", numpy_type_spec),
                ]
            )
            # Data type that encapsulates the entire array plus header and item size info
            dump_dtype = np.dtype(
                [
                    ("n_dim", ">u4"),
                    ("hasnull", ">u4"),
                    ("eltype", ">u4"),
                    ("size_index", ">u4", (len(obj.shape), 2)),
                    ("array", sized_item_dtype, (n_items,)),
                ]
            )
            return np.array(
                [
                    (
                        len(obj.shape),
                        0,  # no nulls
                        element_oid,
                        # Each axis has "dim" elements, starting
                        # at index 1 (in postgresql 1-based indexing)
                        [(dim, 1) for dim in obj.shape],
                        [(item_size, item) for item in obj.flatten()],
                    )
                ],
                dtype=dump_dtype,
            ).tobytes()

    return TypeSpecificDumper(np.ndarray)


@pytest.mark.parametrize("values", (
    [1,2,3,4,5],
    [[1,2,3,4,5],[6,7,8,9,10]],
    [[1,2,3],[4,5,6],[7,8,9]],
))
@pytest.mark.parametrize("array_type",  list(SUPPORTED_NUMPY_TYPES))
def test_array_dumper(values, array_type):
    arr_in = np.array(values, dtype=array_type)
    dumper = ndarray_binary_dumper(array_type)
    buf = dumper.dump(arr_in)
    loader = ArrayBinaryLoader(dumper.oid)
    arr_out = loader.load(buf)
    assert np.allclose(arr_in, arr_out)
```

**Comment by j-carson** [comment]
Definitely hit a few more corner cases, but I have it working - performance wise, it seems to help if you are loading a large array. Thanks for the help!

**Comment by klausfrandsen** [comment]
Came past this thread in my quest to trying to speed up a large SELECT WHERE query and transfer to numpy for further processing. My biggest problem was the (huge) python overhead in terms of the datatypes, as my data was simply millions of rows with four fixed width columns.  

I ended up with a speedup of about a factor of 5x by using this method:
1) CREATE TEMPORARY TABLE tmp_table AS SELECT ... WHERE ...
2) cursor.copy_expert('COPY tmp_table TO STDOUT WITH (FORMAT binary)', file_like)

where file_like is an instance of io.BytesIO()

Then after a bit of seek and trunk to handle the fileheader (see Postgres docs on COPY TO), the binary data can be transferred directly to a numpy variable of composite dtype via numpy.frombuffer()

My measurements show that the previous implementation in python used around 1400 bytes per row fetched, whereas the solution above uses 42 bytes per row fetched. The actual data per row is 24 bytes.

**Comment by j-carson** [comment]
@klausfrandsen  copy_expert is psycopg2, though. Psycopg is generally more memory efficient, the API is simpler and easier to use, and you can use the "copy" object to copy a query without creating a temp table. 

https://www.psycopg.org/psycopg3/docs/basic/copy.html#using-copy-to-and-copy-from

**Comment by jonnor** [comment]
I am interested in this, for the same reasons as described by @klausfrandsen . Since this is about getting data out of postgres, I guess the required extension would be a `Loader`? And not a `Dumper` (for which there is an example above).

**Comment by j-carson** [comment]
@jonnor You are correct that the snippet above is for getting data that's already in a numpy format and putting it into postgres.

 If you want to get data out of postgres and convert it to numpy, I think you want to look at writing a row factory that does the converstion, rather than a dumper. 

https://www.psycopg.org/psycopg3/docs/advanced/rows.html

**Comment by dvarrazzo** [comment]
@jonnor 
> I guess the required extension would be a `Loader`? And not a `Dumper` (for which there is an example above).

Yes, that's correct: a loader would convert a data type such as a float array into a Python object such as a numpy array.

@j-carson 
> If you want to get data out of postgres and convert it to numpy, I think you want to look at writing a row factory that does the converstion, rather than a dumper.

I guess it depends but I'm not sure it matches well this use case. Loaders convert postgres types to python objects. A row factory can then convert a tuple of such Python objects into a different container, but you still end up with a list of such objects (if you fetch them using `fetchall()`.

**Comment by Delengowski** [comment]
I would very much be interested in this request and believe it's best way to solve #316 . Pandas stores data in a combination of numpy/arrow internally and placing the data directly into numpy arrays is probably the fastest way to get it into pandas. 

Getting a query result into a dictionary of numpy arrays is probably ideal

**Comment by Delengowski** [comment]
I'm highly interested in making an attempt here.

I have questions about what the exact requirements for this are.

What I would like is for the cursor to have a method like 

```python
def fetch_numpy(self, *, layout: Literal["columns", "rows"] = "rows") -> dict[str, np.NDArray] | np.NDArray]:
    """Returns a dictionary of Nx1 arrays representing columns or Nx1 array of volatile dtype representing records."""
    ...
```

I believe the most performant case would be doing something in Cython that takes the binary result, does whatever endianess manipulations required and placing them into a buffer to pass to `np.frombuffer()`.

Some of my questions are

1.) Does this seem reasonable?
2.) How do we handle postgres types that do not map nicely into numpy dtypes? For example, an int array type from postgres can become a numpy array and be placed into an array of `object` dtype.
3.) Do we have to support dumping for this PR to be accepted? If so I have concerns with accurately discerning the dtype for more awkward situations. Like dumping an object dtype numpy array, which has numpy int arrays.

**Comment by dvarrazzo** [comment]
> What I would like is for the cursor to have a method like
> ```python
> def fetch_numpy(self, *, layout: Literal["columns", "rows"] = "rows") -> dict[str, np.NDArray] | np.NDArray]:
>     """Returns a dictionary of Nx1 arrays representing columns or Nx1 array of volatile dtype representing records."""
>     ...
> ```

Not at all, no. There's no such similar method in Psycopg, it's not how it would work. The support would be dumpers/loaders to convert between Python numpy and PostgreSQL array, unless there is evidence in the wild that different practices are widely supported (such as dumping numpy arrays to a bytea binary blob).

> 1.) Does this seem reasonable? 

No, I expect whatever support to be part of the adapters infrastructure.

> 2.) How do we handle postgres types that do not map nicely into numpy dtypes? For example, an int array type from postgres can become a numpy array and be placed into an array of `object` dtype. 

Can you propose a dtype <-> postgres types conversion grid? I understand int16, int32, int64, float32, float64, bool arrays map well between numpy and postgres, other types not so (such as Python unsigned ones or Postgres numeric)

> 3.) Do we have to support dumping for this PR to be accepted? If so I have concerns with accurately discerning the dtype for more awkward situations. Like dumping an object dtype numpy array, which has numpy int arrays.

I would say that supporting loading but not dumping makes little sense. However we don't have to support absolutely everything, for example we don't have to dump anything that then we cannot load.

**Comment by Delengowski** [comment]
First, my desire is to get data into a pandas dataframe as fast possible. For that I think a dict of numpy arrays is the ideal case. 

With that said here's a table I made, I tried to keep in mind what psycopg currently does for postgres types.

| PostgreSQL Type | NumPy dtype | Notes |
|---|---|---|
| smallint (int2) | np.int16 | Direct mapping |
| integer (int4) | np.int32 | Direct mapping |
| bigint (int8) | np.int64 | Direct mapping |
| serial | np.int32 | Underlying type is int4 |
| bigserial | np.int64 | Underlying type is int8 |
| real (float4) | np.float32 | Direct mapping |
| double precision (float8) | np.float64 | Direct mapping |
| numeric / decimal | np.object_ | python Decimal objects |
| boolean | np.bool_ | Direct mapping |
| char(n) | np.bytes_ or np.str_ | Fixed-length string. np.bytes_ aligns better but I'm more in favor of np.str_ |
| varchar(n) | np.str_ | Variable-length string |
| text | np.str_ | Variable-length UTF-8 |
| bytea | np.bytes_ or np.object_ | Raw binary |
| date | np.datetime64[D] | Day precision |
| timestamp | np.datetime64[us] | Microsecond precision typical |
| timestamptz | np.datetime64[us] | Timezone lost |
| time | np.timedelta64[us] | Duration since midnight |
| interval | np.timedelta64[us] | Approximation (months/years ambiguous) |
| uuid | np.object_ | Python uuid.UUID |
| json / jsonb | np.object_ | Python dict |
| array | np.object_ | numpy array holding numpy arrays. We could concatenate if they're all same dimesnions but that seems like a lot of work |
| hstore | np.object_ | Python dict |
| composite types | np.void | structs, would requires equivalent TYPE to be present in postgres |
| range types | np.object_ | psycopg.Range |
| xml | np.object_ | python str |
| N/A | np.complex* | not mappable |
| N/A | np.uint* | not mappable |
| N/A | np.float16/128 | not mappable |


For loading from postgres, I think we can have a relatively easy time. Dumping is another beast. For the number of types that are using np.object_, we will have to manually inspect element to deduce the type. How we can do that reliably without inspecting the entire array is troublesome for me. Unless we require the mapped datatypes to be specified by the caller.

**Comment by dvarrazzo** [comment]
@Delengowski 

> First, my desire is to get data into a pandas dataframe as fast possible.

From where: from a PostgreSQL array type or from a normal query?

This ticket is about mapping Postgres array types such as `int8[]` into numpy instances and back. If you want to run a generic query and find your data in a dataframe then this requires an entirely different adaptation system and is a matter that was discussed in #316

**Comment by Delengowski** [comment]
Okay. I see the disconnect I'm having now. I misunderstood this issue and one linked on OP. 

I believed this was more about getting rows of data into a numpy arrays not loading a postgres array to numpy array.

To answer your question, yes I meant a normal query.

**Comment by Delengowski** [comment]
There's a comment on #316 that links to this issue, which is where I believe my misunderstanding began.

**Comment by dvarrazzo** [comment]
@Delengowski I understand the confusion, no problem. I think there are use cases for wanting to adapt Postgres arrays back and forth numpy arrays (not a really strong one, testament the little work that this issue has seen) and converting a whole query result to a memory-efficient structure rather than a list of Python objects as the current infrastructure allows.
